private/Metaprogram.h

```diff
@@ -0,0 +1,349 @@
+#pragma once
+
+#include <tuple>
+#include <assert.h>
+
+#include "TypeList.h"
+#include "TupleManipulation.h"
+
+#define PANKU_LIST(...) \
+   using PankuUserProvidedList = type_list<__VA_ARGS__>; \
+   using PankuProcessedList = PankuMetaprogram::type_list_extract_heads<PankuUserProvidedList>; \
+   using PankuUserClassList = PankuProcessedList::type; \
+   using PankuDependencies = PankuProcessedList::headless_type; \
+   using PankuPath = PankuMetaprogram::topological_sort<PankuUserClassList, PankuDependencies>::type; \
+   using PankuPathPointerised = PankuMetaprogram::type_list_pointerise<PankuPath>::type; \
+   using PankuClassTuple = convert<PankuPathPointerised, std::tuple>; \
+   extern Panku<PankuClassTuple> panku;
+
+#define PANKU_INSTANCE Panku<PankuClassTuple> panku;
+
+#define STANDALONE(ClassName) type_list<ClassName>
+#define DEPENDENCY(ClassName, ...) type_list<ClassName, ##__VA_ARGS__>
+
+using namespace TypeList;
+
+template<typename UserClass>
+UserClass& ConstructAndInitialise();
+
+template<typename Tuple>
+class Panku
+{
+   public:
+   Panku() 
+   {
+      TupleManipulation::for_each_in_tuple(userClassTuple, [](auto& element)
+            { element = &ConstructAndInitialise<decltype(*element)>(); });
+   }
+
+   template<typename UserClass>
+   UserClass& Get() 
+   {
+      auto userObjectPointer = std::get<UserClass*>(userClassTuple);
```

Comment:
The tuple needs to be a tuple of pointers because they are nullable, I need them to start at void so I can build up the user classes incrementally through the user supplied "ConstructAndInitialise" functions.

I tried to store a tuple of concretes here but it's not really possible unless every object has a trivial constructor :)